src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -440,7 +481,11 @@ public void DoWriteIfNeeded()
                 }, this);
 
                 Self._head = _lockedEnd.Block;
-                Self._head.Start = _lockedEnd.Index;
```

Comment:
Nice catch.

If you just want to avoid the NullRefEx, you should do the following:

    var newHead = _lockedEnd.Block;
    newHead.Start = _lockedEnd.Index;

    if (Self._head != null)
    {
        Self._head = newHead;
    }

With the current change, you could still get a null ref if `_head` is set to null between the if comparison and the actual assignment to `Self._head.Start`. This makes the race a lot tighter than it was before, but it's still there.

To be safe, all modifications to _head should be made with the `_returnLock`. This fixes the race because `_head` is only set to null when the `_returnLock` is taken.